Explore: show meta data when available

besartberisha opened the pull request:
This PR is responsible for showing `metaData` when those data are available. 

In order to properly show the metaData I needed to check in multiple places if the data now that isn't available since the source of data and metaData is in one place.

Fixes #25515

besartberisha committed:
Return valid metadata even if data is empty

besartberisha committed:
Make sure fields is not accessed when there is no data

besartberisha committed:
Meta Info should depend only on meta data

tskarhed commented on packages/grafana-data/src/dataframe/processDataFrame.ts:
```diff
@@ -248,22 +248,24 @@ export function guessFieldTypeForField(field: Field): FieldType | undefined {
-          };
-        }),
-      };
+  if (series.fields) {
```
Maybe do an early return here instead. Much more readable.

```ts
if(!series.fields){
  return series;
}
```

besartberisha replied:
Yep I agree, changed it to make it more readable. Thank you.

besartberisha committed:
Refactor condition for better reading

besartberisha committed:
Change data to empty array for better condition handling

besartberisha commented:
I have updated the PR, and added additional checks to make sure that empty fields aren't passed down and the proper meta data is shown when there is meta data.

ivanahuckova commented on public/app/plugins/datasource/loki/result_transformer.ts:
```diff
@@ -301,21 +301,36 @@ export function lokiStreamsToDataframes(
+    stats,
+    custom,
+  };
+
```
I really like that you moved this from return . 💙

ivanahuckova reviewed (approved):
Looks good to me! I would also wait for someone else to review this changes as well. 🙂

zoltanbedi reviewed (approved):
lgtm 🥓

davkal commented:
Let's wait for unification to be merged.

besartberisha committed:
Merge remote-tracking branch 'origin/master' into besart/feature-show-processed-logs

kaydelaney commented on packages/grafana-data/src/dataframe/processDataFrame.ts:
```diff
@@ -251,6 +251,10 @@ export function guessFieldTypeForField(field: Field): FieldType | undefined {
  * @param guessDefined Whether to guess types of fields with already defined types
  */
 export const guessFieldTypes = (series: DataFrame, guessDefined = false): DataFrame => {
+  if (!series.fields.length) {
```
Is this if statement necessary? If the dataframe `series` has no fields, then the for loop at 258 should just be skipped and then the series will be returned at line 277 I think

besartberisha replied:
Good catch, this was necessary before since fields were undefined but now it's an empty array.

kaydelaney commented on public/app/core/logs_model.ts:
```diff
@@ -186,7 +186,11 @@ export function makeSeriesForLogs(sortedRows: LogRowModel[], bucketSize: number,
 function isLogsData(series: DataFrame) {
-  return series.fields.some(f => f.type === FieldType.time) && series.fields.some(f => f.type === FieldType.string);
+  return (
+    series.fields.length &&
```
Also not sure that `series.fields.length` is necessary here. If `series.fields` is of length 0 then the following `series.fields.some(...)` expressions should evaluate to false

besartberisha replied:
Not needed, same as the comment above I've removed it.

kaydelaney commented on public/app/core/logs_model.ts:
```diff
@@ -445,6 +459,15 @@ export function logSeriesToLogsModel(logSeries: DataFrame[]): LogsModel | undefi
 }
 
+function hasFields(logSeries: DataFrame[]): boolean {
+  for (const series of logSeries) {
```
i think this function could be simplified by just returning
`logSeries.some(series => series.fields.length)`

kaydelaney commented on public/app/core/logs_model.ts:
```diff
@@ -303,30 +307,40 @@ export function logSeriesToLogsModel(logSeries: DataFrame[]): LogsModel | undefi
   let hasUniqueLabels = false;
 
   for (const info of allSeries) {
+    if (!info) {
```
Can `info` be undefined?

besartberisha replied:
No, I've removed the check since it was part of the issue similar to the other comment.

besartberisha committed:
Remove unecessary checks

aocenas commented:
@besartberisha can we also add/update tests for this?

besartberisha committed:
Add test for showing metaData even when data is empty
